design/src/aws_config.md

```diff
@@ -0,0 +1,161 @@
+        todo!()
+    }
+
+    pub fn region(&self) -> Option<Region> {
+        todo!()
+    }
+
+    pub fn sleep(&self) -> impl Sleep {
+        todo!()
+    }
+
+    pub fn retry_config(&self) ->...// eg. mode, max attempts etc.
+
+    pub fn timeout_config(&self) -> // eg. request timeout, read timeout, etc.
+}
+```
+
+The `Builder` for `SharedConfig` allows customers to provide individual overrides and handles the insertion of the
+default chain for regions, sleep, connectors, and credentials.
+
+## Sleep + Connector
+
+Sleep and Connector are both runtime dependent features. `aws-config` will use `tokio` and `hyper` as optional
+features. **This centralizes the Tokio/Hyper dependency** removing the need for each service to maintain their own
+Tokio/Hyper features.
+
+## The `.build()` method on service config
+
+Currently, the `.build()` method on service config will fill in defaults. As part of this change, `.build()` called on
+the service config with missing properties would be a runtime panic. Most customers would utilize the `From`
+implementation.
+
+## Stability and Versioning
+
+The introduction of `SharedConfig` to aws-types is not without risks. If a customer depends on a version aws-config that
+uses `SharedConfig` that is incompatible, they will get confusing compiler errors.
+
+An example of a problematic version set:
+
+```
+┌─────────────────┐                 ┌───────────────┐
+│ aws-types = 0.1 │                 │aws-types= 0.2 │
+└─────────────────┘                 └───────────────┘
+           ▲                                 ▲
+           │                                 │
+           │                                 │
+           │                                 │
+ ┌─────────┴─────────────┐          ┌────────┴───────┐
+ │aws-sdk-dynamodb = 0.5 │          │aws-config = 0.6│
+ └───────────┬───────────┘          └───────┬────────┘
+             │                              │
+             │                              │
+             │                              │
+             │                              │
+             │                              │
+             ├─────────────────────┬────────┘
+             │ my-lambda-function  │
+             └─────────────────────┘
+```
+
+To mitigate this risk, we will need to make `aws-types` essentially permanently stable. Changes
+to `aws-types` need to be made with extreme care. This will ensure that two versions of `aws-types` never end up in a customer's dependency tree.
+
+We will dramatically reduce the surface area of `aws-types` to contain only interfaces.
+
+Several breaking changes will be made as part of this, notably, the profile file parsing will be moved out of aws-types.
+
+## Changes Checklist
+
+- [ ] ProvideRegion becomes async using a newtype'd future.
+- [ ] AsyncProvideCredentials is removed. ProvideCredentials becomes async using a newtype'd future.
+- [ ] ProvideCredentials moved into `aws-types`. `Credentials` moved into `aws-types`
```

Comment:
Do we want `ProvideCredentials` to be permanently stable before `async trait` is available?

Reply:
hmm...not much choice unfortunately. I think this will probably be something we have to take care of between 0.1 and 1.0